Automatically wrap autograd.grad with allow_in_graph

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #172126
* #169415
* __->__ #169493

In this PR, we add support for tracing through autograd.grad. Some changelogs: 

- Add trace_autograd_ops config (default False)
- Add handler for autograd.grad that checks for external grad_fn inputs
- Add validation for compile inputs/outputs:
   - Graph break when autograd.grad inputs are reachable to any inputs that have external grad fn, this is because torch.compile won't understand the external relations outside compiled regions 
   - Graph break when compile outputs are reachable from autograd.grad outputs as we will run into awkward double backward error. 
- Add test_fwd_loss_bwd.py with tests for the new functionality
- Graph break on GradientEdge. 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo

tugsbayasgalan committed:
Automatically wrap autograd.grad with allow_in_graph

Add trace_autograd_ops config to control whether torch.autograd.grad
and torch.autograd.backward are traced or cause graph breaks.

- Add trace_autograd_ops config (default True)
- Add handler for autograd.grad that checks for external grad_fn inputs
- Add validation for outputs requiring grad when autograd.grad is used
- Add test_fwd_loss_bwd.py with tests for the new functionality

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Automatically wrap autograd.grad with allow_in_graph"


Add trace_autograd_ops config to control whether torch.autograd.grad
and torch.autograd.backward are traced or cause graph breaks.

- Add trace_autograd_ops config (default True)
- Add handler for autograd.grad that checks for external grad_fn inputs
- Add validation for outputs requiring grad when autograd.grad is used
- Add test_fwd_loss_bwd.py with tests for the new functionality

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Automatically wrap autograd.grad with allow_in_graph"


Add trace_autograd_ops config to control whether torch.autograd.grad
and torch.autograd.backward are traced or cause graph breaks.

- Add trace_autograd_ops config (default True)
- Add handler for autograd.grad that checks for external grad_fn inputs
- Add validation for outputs requiring grad when autograd.grad is used
- Add test_fwd_loss_bwd.py with tests for the new functionality

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Automatically wrap autograd.grad with allow_in_graph"


Add trace_autograd_ops config to control whether torch.autograd.grad
and torch.autograd.backward are traced or cause graph breaks.

- Add trace_autograd_ops config (default True)
- Add handler for autograd.grad that checks for external grad_fn inputs
- Add validation for outputs requiring grad when autograd.grad is used
- Add test_fwd_loss_bwd.py with tests for the new functionality

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Automatically wrap autograd.grad with allow_in_graph"


Add trace_autograd_ops config to control whether torch.autograd.grad
and torch.autograd.backward are traced or cause graph breaks.

- Add trace_autograd_ops config (default True)
- Add handler for autograd.grad that checks for external grad_fn inputs
- Add validation for outputs requiring grad when autograd.grad is used
- Add test_fwd_loss_bwd.py with tests for the new functionality

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Automatically wrap autograd.grad with allow_in_graph"


Add trace_autograd_ops config to control whether torch.autograd.grad
and torch.autograd.backward are traced or cause graph breaks.

- Add trace_autograd_ops config (default True)
- Add handler for autograd.grad that checks for external grad_fn inputs
- Add validation for outputs requiring grad when autograd.grad is used
- Add test_fwd_loss_bwd.py with tests for the new functionality

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Automatically wrap autograd.grad with allow_in_graph"


Add trace_autograd_ops config to control whether torch.autograd.grad
and torch.autograd.backward are traced or cause graph breaks.

- Add trace_autograd_ops config (default True)
- Add handler for autograd.grad that checks for external grad_fn inputs
- Add validation for outputs requiring grad when autograd.grad is used
- Add test_fwd_loss_bwd.py with tests for the new functionality

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Automatically wrap autograd.grad with allow_in_graph"


Add trace_autograd_ops config to control whether torch.autograd.grad
and torch.autograd.backward are traced or cause graph breaks.

- Add trace_autograd_ops config (default True)
- Add handler for autograd.grad that checks for external grad_fn inputs
- Add validation for outputs requiring grad when autograd.grad is used
- Add test_fwd_loss_bwd.py with tests for the new functionality

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan commented:
@pytorchbot label "topic: not user facing"

tugsbayasgalan committed:
Update on "Automatically wrap autograd.grad with allow_in_graph"


Add trace_autograd_ops config to control whether torch.autograd.grad
and torch.autograd.backward are traced or cause graph breaks.

- Add trace_autograd_ops config (default True)
- Add handler for autograd.grad that checks for external grad_fn inputs
- Add validation for outputs requiring grad when autograd.grad is used
- Add test_fwd_loss_bwd.py with tests for the new functionality

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

ezyang commented on torch/_dynamo/output_graph.py:
```diff
@@ -677,6 +677,15 @@ def __init__(
+        # Tracks if torch.autograd.grad is used in the graph. This is set during
+        # tracing when we encounter a call to torch.autograd.grad, and is used
+        # to validate that graph inputs are leaf tensors (no external grad_fn).
+        self.uses_autograd_grad = False
```
Oh, this is for the requires grad outputs test. OK.

ezyang replied:
I think I am still confused though. In `_validate_autograd_grad_outputs` you are only checking the outputs of the ENTIRE dynamo graph, but what I care about is the outputs of the torch.autograd.grad, because if torch.autograd.grad outputs don't require gradients, it actually should work for AOTAutograd to do the global backwards (since there's some disjoint thing going on)

tugsbayasgalan replied:
https://github.com/pytorch/pytorch/blob/3c551c2480fa4d38f3fe634a39e7033a29a2517e/torch/_functorch/aot_autograd.py#L586 this is to not trip at this check.

ezyang commented on torch/_dynamo/variables/torch.py:
```diff
@@ -1444,6 +1444,86 @@ def handle_check(self, tx: "InstructionTranslator", *args, **kwargs):
+                            "receives tensors with grad_fn created outside the compiled region. "
+                            "The autograd graph cannot extend beyond the compiled region boundary. "
+                            "Note: This is a conservative check. It may be safe if the input with "
+                            "grad_fn is not reachable from the autograd graph of the differentiated tensors."
```
Honestly, let's just do this properly with a `grad_fn` traversal, I think...

tugsbayasgalan replied:
Yep done.

tugsbayasgalan commented on torch/_dynamo/variables/torch.py:
```diff
@@ -1444,6 +1444,86 @@ def handle_check(self, tx: "InstructionTranslator", *args, **kwargs):
+                        )
+                elif isinstance(var, BaseListVariable):
+                    for item in var.items:
+                        check_lazy_tensor_grad_fn(item)
```
Yeah i think this is probably fine (All tuple and lists inherit from BaseListVariable) and I think GradientEdge class in autograd is not supported in dynamo today so it will graph break early. But I will add extra checks to just to be safe.

tugsbayasgalan committed:
Update on "Automatically wrap autograd.grad with allow_in_graph"


Add trace_autograd_ops config to control whether torch.autograd.grad
and torch.autograd.backward are traced or cause graph breaks.

- Add trace_autograd_ops config (default True)
- Add handler for autograd.grad that checks for external grad_fn inputs
- Add validation for outputs requiring grad when autograd.grad is used
- Add test_fwd_loss_bwd.py with tests for the new functionality

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Automatically wrap autograd.grad with allow_in_graph"


Add trace_autograd_ops config to control whether torch.autograd.grad
and torch.autograd.backward are traced or cause graph breaks.

- Add trace_autograd_ops config (default False)
- Add handler for autograd.grad that checks for external grad_fn inputs
- Add validation for compile inputs/outputs requiring grad when autograd.grad is used. We do dfs on inputs/outputs to make sure they are not reachable from autograd.grad IF they have external requires_grad fn. If they have requires_grad fn and is connected to autograd.grad, we graph break. 
- Add test_fwd_loss_bwd.py with tests for the new functionality

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Automatically wrap autograd.grad with allow_in_graph"


Add trace_autograd_ops config to control whether torch.autograd.grad
and torch.autograd.backward are traced or cause graph breaks.

- Add trace_autograd_ops config (default False)
- Add handler for autograd.grad that checks for external grad_fn inputs
- Add validation for compile inputs/outputs requiring grad when autograd.grad is used. We do dfs on inputs/outputs to make sure they are not reachable from autograd.grad IF they have external requires_grad fn. If they have requires_grad fn and is connected to autograd.grad, we graph break. 
- Add test_fwd_loss_bwd.py with tests for the new functionality

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Automatically wrap autograd.grad with allow_in_graph"


Add trace_autograd_ops config to control whether torch.autograd.grad
and torch.autograd.backward are traced or cause graph breaks.

- Add trace_autograd_ops config (default False)
- Add handler for autograd.grad that checks for external grad_fn inputs
- Add validation for compile inputs/outputs requiring grad when autograd.grad is used. We do dfs on inputs/outputs to make sure they are not reachable from autograd.grad IF they have external requires_grad fn. If they have requires_grad fn and is connected to autograd.grad, we graph break. 
- Add test_fwd_loss_bwd.py with tests for the new functionality

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Automatically wrap autograd.grad with allow_in_graph"


Add trace_autograd_ops config to control whether torch.autograd.grad
and torch.autograd.backward are traced or cause graph breaks.

- Add trace_autograd_ops config (default False)
- Add handler for autograd.grad that checks for external grad_fn inputs
- Add validation for compile inputs/outputs requiring grad when autograd.grad is used. We do dfs on inputs/outputs to make sure they are not reachable from autograd.grad IF they have external requires_grad fn. If they have requires_grad fn and is connected to autograd.grad, we graph break. 
- Add test_fwd_loss_bwd.py with tests for the new functionality

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Automatically wrap autograd.grad with allow_in_graph"


Add trace_autograd_ops config to control whether torch.autograd.grad
and torch.autograd.backward are traced or cause graph breaks.

- Add trace_autograd_ops config (default False)
- Add handler for autograd.grad that checks for external grad_fn inputs
- Add validation for compile inputs/outputs requiring grad when autograd.grad is used. We do dfs on inputs/outputs to make sure they are not reachable from autograd.grad IF they have external requires_grad fn. If they have requires_grad fn and is connected to autograd.grad, we graph break. 
- Add test_fwd_loss_bwd.py with tests for the new functionality

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Automatically wrap autograd.grad with allow_in_graph"


Add trace_autograd_ops config to control whether torch.autograd.grad
and torch.autograd.backward are traced or cause graph breaks.

- Add trace_autograd_ops config (default False)
- Add handler for autograd.grad that checks for external grad_fn inputs
- Add validation for compile inputs/outputs requiring grad when autograd.grad is used. We do dfs on inputs/outputs to make sure they are not reachable from autograd.grad IF they have external requires_grad fn. If they have requires_grad fn and is connected to autograd.grad, we graph break. 
- Add test_fwd_loss_bwd.py with tests for the new functionality

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Automatically wrap autograd.grad with allow_in_graph"


Add trace_autograd_ops config to control whether torch.autograd.grad
and torch.autograd.backward are traced or cause graph breaks.

- Add trace_autograd_ops config (default False)
- Add handler for autograd.grad that checks for external grad_fn inputs
- Add validation for compile inputs/outputs requiring grad when autograd.grad is used. We do dfs on inputs/outputs to make sure they are not reachable from autograd.grad IF they have external requires_grad fn. If they have requires_grad fn and is connected to autograd.grad, we graph break. 
- Add test_fwd_loss_bwd.py with tests for the new functionality

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Automatically wrap autograd.grad with allow_in_graph"


In this PR, we add support for tracing through autograd.grad. Some changelogs: 

- Add trace_autograd_ops config (default False)
- Add handler for autograd.grad that checks for external grad_fn inputs
- Add validation for compile inputs/outputs:
   - Graph break when autograd.grad inputs are reachable to any inputs that have external grad fn, this is because torch.compile won't understand the external relations outside compiled regions 
   - Graph break when compile outputs are reachable from autograd.grad outputs as we will run into awkward double backward error. 
- Add test_fwd_loss_bwd.py with tests for the new functionality
- Graph break on GradientEdge. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Automatically wrap autograd.grad with allow_in_graph"


In this PR, we add support for tracing through autograd.grad. Some changelogs: 

- Add trace_autograd_ops config (default False)
- Add handler for autograd.grad that checks for external grad_fn inputs
- Add validation for compile inputs/outputs:
   - Graph break when autograd.grad inputs are reachable to any inputs that have external grad fn, this is because torch.compile won't understand the external relations outside compiled regions 
   - Graph break when compile outputs are reachable from autograd.grad outputs as we will run into awkward double backward error. 
- Add test_fwd_loss_bwd.py with tests for the new functionality
- Graph break on GradientEdge. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Automatically wrap autograd.grad with allow_in_graph"


In this PR, we add support for tracing through autograd.grad. Some changelogs: 

- Add trace_autograd_ops config (default False)
- Add handler for autograd.grad that checks for external grad_fn inputs
- Add validation for compile inputs/outputs:
   - Graph break when autograd.grad inputs are reachable to any inputs that have external grad fn, this is because torch.compile won't understand the external relations outside compiled regions 
   - Graph break when compile outputs are reachable from autograd.grad outputs as we will run into awkward double backward error. 
- Add test_fwd_loss_bwd.py with tests for the new functionality
- Graph break on GradientEdge. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan commented on torch/_dynamo/output_graph.py:
```diff
@@ -2092,6 +2126,58 @@ def run_compiler_collective(self) -> None:
+            if (
+                fake_tensor is not None
+                and find_reachable_grad_fn(
+                    fake_tensor, self.autograd_grad_output_grad_fns
```
Also add this test https://github.com/pytorch/pytorch/pull/169493/files#diff-54f61291a4be8afc39154a2c8c04c91a3fef691f3bdb1b2b8180b55bffbe8448R658 which shows that when we are collecting consumed grad functions, we exclude autograd.grad input grad fn's

tugsbayasgalan committed:
Update on "Automatically wrap autograd.grad with allow_in_graph"


In this PR, we add support for tracing through autograd.grad. Some changelogs: 

- Add trace_autograd_ops config (default False)
- Add handler for autograd.grad that checks for external grad_fn inputs
- Add validation for compile inputs/outputs:
   - Graph break when autograd.grad inputs are reachable to any inputs that have external grad fn, this is because torch.compile won't understand the external relations outside compiled regions 
   - Graph break when compile outputs are reachable from autograd.grad outputs as we will run into awkward double backward error. 
- Add test_fwd_loss_bwd.py with tests for the new functionality
- Graph break on GradientEdge. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Automatically wrap autograd.grad with allow_in_graph"


In this PR, we add support for tracing through autograd.grad. Some changelogs: 

- Add trace_autograd_ops config (default False)
- Add handler for autograd.grad that checks for external grad_fn inputs
- Add validation for compile inputs/outputs:
   - Graph break when autograd.grad inputs are reachable to any inputs that have external grad fn, this is because torch.compile won't understand the external relations outside compiled regions 
   - Graph break when compile outputs are reachable from autograd.grad outputs as we will run into awkward double backward error. 
- Add test_fwd_loss_bwd.py with tests for the new functionality
- Graph break on GradientEdge. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Automatically wrap autograd.grad with allow_in_graph"


In this PR, we add support for tracing through autograd.grad. Some changelogs: 

- Add trace_autograd_ops config (default False)
- Add handler for autograd.grad that checks for external grad_fn inputs
- Add validation for compile inputs/outputs:
   - Graph break when autograd.grad inputs are reachable to any inputs that have external grad fn, this is because torch.compile won't understand the external relations outside compiled regions 
   - Graph break when compile outputs are reachable from autograd.grad outputs as we will run into awkward double backward error. 
- Add test_fwd_loss_bwd.py with tests for the new functionality
- Graph break on GradientEdge. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

ezyang commented on torch/_dynamo/variables/torch.py:
```diff
@@ -1505,6 +1566,236 @@ def handle_check(self, tx: "InstructionTranslator", *args, **kwargs):
+            3. Reject external GradientEdge objects (can't trace external autograd nodes)
+            4. Collect external grad_fns from graph inputs
+            5. Compute "consumed" grad_fns: all nodes reachable from outputs, stopping at inputs
+            6. Validate: no external grad_fn (except inputs) is in consumed set
```
Yes so specifically idk what you mean by (inputs) here. Because when we fakeify, we only fakeify inputs, we don't fakeify beyond them (and the fakeified grad_fn carrying inputs end up with fake grad_fns from fakeification)

tugsbayasgalan replied:
Sorry i think the confusion is coming from the fact that there are two kind of inputs we are talking about here:
1. Graph inputs: tensors passed into the torch.compile region from outside
2. autograd.grad inputs: the inputs parameter to torch.autograd.grad(outputs, inputs)

inputs here refers to the inputs parameter to autograd.grad function call. In this case, we don't care because autograd.grad won't consume the grad functions of input parameter.

tugsbayasgalan committed:
Update on "Automatically wrap autograd.grad with allow_in_graph"


In this PR, we add support for tracing through autograd.grad. Some changelogs: 

- Add trace_autograd_ops config (default False)
- Add handler for autograd.grad that checks for external grad_fn inputs
- Add validation for compile inputs/outputs:
   - Graph break when autograd.grad inputs are reachable to any inputs that have external grad fn, this is because torch.compile won't understand the external relations outside compiled regions 
   - Graph break when compile outputs are reachable from autograd.grad outputs as we will run into awkward double backward error. 
- Add test_fwd_loss_bwd.py with tests for the new functionality
- Graph break on GradientEdge. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Automatically wrap autograd.grad with allow_in_graph"


In this PR, we add support for tracing through autograd.grad. Some changelogs: 

- Add trace_autograd_ops config (default False)
- Add handler for autograd.grad that checks for external grad_fn inputs
- Add validation for compile inputs/outputs:
   - Graph break when autograd.grad inputs are reachable to any inputs that have external grad fn, this is because torch.compile won't understand the external relations outside compiled regions 
   - Graph break when compile outputs are reachable from autograd.grad outputs as we will run into awkward double backward error. 
- Add test_fwd_loss_bwd.py with tests for the new functionality
- Graph break on GradientEdge. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

ezyang commented on torch/_dynamo/variables/torch.py:
```diff
@@ -1505,6 +1566,240 @@ def handle_check(self, tx: "InstructionTranslator", *args, **kwargs):
+            inputs_grad_fns: set[torch.autograd.graph.Node] = set()
+            for tensor, _ in inputs_with_sources:
+                if isinstance(tensor, torch.Tensor) and tensor.grad_fn is not None:
+                    inputs_grad_fns.add(tensor.grad_fn)
```
The tensor.grad_fn test here is to account for leafs, but at this point in time it's not obvious how leaf inputs are going to be handled (leaf here is requires_grad=True, grad_fn=None)

ezyang commented on torch/_dynamo/variables/torch.py:
```diff
@@ -1505,6 +1566,240 @@ def handle_check(self, tx: "InstructionTranslator", *args, **kwargs):
+            external_grad_fns: set[torch.autograd.graph.Node] = set()
+            # Map grad_fn -> source name for better error messages
+            grad_fn_to_source: dict[torch.autograd.graph.Node, str] = {}
+            for var in tx.output.input_source_to_var.values():
```
We need to more exhaustively handle ALL grad_fn associated with input tensors. There are some more cases that aren't covered here: (1) if I have a view, I need the grad_fn of the base (alternately, I can recursively traverse all reachable grad_fns from the main tensors); (2) if I have a leaf, I need its implicit accumulator grad fn (the AcccumulateGrad one) which I can get at by doing leaf.view(-1).grad_fn.next_functions[0].grad_fn; (3) tensor subclasses (it should be unusual for the inner tensors to require grad, and DTensor will not have this, but we do need to consider it).

tugsbayasgalan committed:
Update on "Automatically wrap autograd.grad with allow_in_graph"


In this PR, we add support for tracing through autograd.grad. Some changelogs: 

- Add trace_autograd_ops config (default False)
- Add handler for autograd.grad that checks for external grad_fn inputs
- Add validation for compile inputs/outputs:
   - Graph break when autograd.grad inputs are reachable to any inputs that have external grad fn, this is because torch.compile won't understand the external relations outside compiled regions 
   - Graph break when compile outputs are reachable from autograd.grad outputs as we will run into awkward double backward error. 
- Add test_fwd_loss_bwd.py with tests for the new functionality
- Graph break on GradientEdge. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Automatically wrap autograd.grad with allow_in_graph"


In this PR, we add support for tracing through autograd.grad. Some changelogs: 

- Add trace_autograd_ops config (default False)
- Add handler for autograd.grad that checks for external grad_fn inputs
- Add validation for compile inputs/outputs:
   - Graph break when autograd.grad inputs are reachable to any inputs that have external grad fn, this is because torch.compile won't understand the external relations outside compiled regions 
   - Graph break when compile outputs are reachable from autograd.grad outputs as we will run into awkward double backward error. 
- Add test_fwd_loss_bwd.py with tests for the new functionality
- Graph break on GradientEdge. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Automatically wrap autograd.grad with allow_in_graph"


In this PR, we add support for tracing through autograd.grad. Some changelogs: 

- Add trace_autograd_ops config (default False)
- Add handler for autograd.grad that checks for external grad_fn inputs
- Add validation for compile inputs/outputs:
   - Graph break when autograd.grad inputs are reachable to any inputs that have external grad fn, this is because torch.compile won't understand the external relations outside compiled regions 
   - Graph break when compile outputs are reachable from autograd.grad outputs as we will run into awkward double backward error. 
- Add test_fwd_loss_bwd.py with tests for the new functionality
- Graph break on GradientEdge. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Automatically wrap autograd.grad with allow_in_graph"


In this PR, we add support for tracing through autograd.grad. Some changelogs: 

- Add trace_autograd_ops config (default False)
- Add handler for autograd.grad that checks for external grad_fn inputs
- Add validation for compile inputs/outputs:
   - Graph break when autograd.grad inputs are reachable to any inputs that have external grad fn, this is because torch.compile won't understand the external relations outside compiled regions 
   - Graph break when compile outputs are reachable from autograd.grad outputs as we will run into awkward double backward error. 
- Add test_fwd_loss_bwd.py with tests for the new functionality
- Graph break on GradientEdge. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Automatically wrap autograd.grad with allow_in_graph"


In this PR, we add support for tracing through autograd.grad. Some changelogs: 

- Add trace_autograd_ops config (default False)
- Add handler for autograd.grad that checks for external grad_fn inputs
- Add validation for compile inputs/outputs:
   - Graph break when autograd.grad inputs are reachable to any inputs that have external grad fn, this is because torch.compile won't understand the external relations outside compiled regions 
   - Graph break when compile outputs are reachable from autograd.grad outputs as we will run into awkward double backward error. 
- Add test_fwd_loss_bwd.py with tests for the new functionality
- Graph break on GradientEdge. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]
